sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+# Shared DPC++ libraries
+
+This document describes purpose and design of Shared DPC++ libraries feature.
+
+## Background
+Sometimes users want to provide *device* functions via shared libraries.
+Simple source example:
+```
+// App:
+
+CGH.parallel_for<app_kernel>(/* ... */ {
+  library_function();
+});
+
+
+// Shared library:
+SYCL_EXTERNAL void library_function() {
+  // do something
+}
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+The main purpose of this feature is to provide a mechanism which allows to
+provide *device* functions via shared libraries and works as close as possible
+to regular shared libraries.
+
+## Requrements:
+User's code is compiled into a shared library which consists of some host API,
+device code and device API (`SYCL_EXTERNAL` functions). The library is linked to
+a user's application which also contains some device code and performs
+computations using DPC++/SYCL.
```

Comment:
Please also support the scenario above when there is no shared library (as described above)

Reply:
Ok.